### ISSUE TITLE
sourceText.startsWith is not a function (v4.6.0)

### ISSUE THREAD

**Issue by SimulatedGREG** [eslint/eslint#9204]
**Tell us about your environment**

* **ESLint Version:** 4.6.0
* **Node Version:** 8.4.0
* **npm Version:** 5.3.0

**What parser (default, Babel-ESLint, etc.) are you using?**
`babel-eslint`

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
  root: true,
  parser: 'babel-eslint',
  parserOptions: {
    sourceType: 'module'
  },
  env: {
    browser: true,
    node: true
  },
  globals: {
    __static: true
  },
  plugins: [
    'html'
  ],
  'rules': {
    'no-debugger': process.env.NODE_ENV === 'production' ? 2 : 0
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue.**
Using the [electron-vue](https://github.com/SimulatedGREG/electron-vue) boilerplate, default scaffold with eslint enabled using an empty configuration (also tried, using `eslint-config-standard` and `eslint-config-airbnb-base`).

**What did you expect to happen?**
Expected code to be linted and *fixed* when possible. It has worked without error [beforehand](https://semaphoreci.com/simulatedgreg/electron-vue/branches/master/builds/326), but the new 4.6.0 version [fails](https://semaphoreci.com/simulatedgreg/electron-vue/branches/master/builds/327) and produces the below error log. Reverting back to 4.5.0 does not produce the error and works as expected.

```bash
# original command, removing formatter doesn't fix either
eslint --ext .js,.vue -f ./node_modules/eslint-friendly-formatter --fix src
```

**What actually happened? Please include the actual, raw output from ESLint.**

```bash
> j@0.0.0 lint:fix /Users/gregholguin/Documents/Projects/playground/j
> eslint --ext .js,.vue -f ./node_modules/eslint-friendly-formatter --fix src

sourceText.startsWith is not a function
TypeError: sourceText.startsWith is not a function
    at Function.SourceCodeFixer.applyFixes (/Users/gregholguin/Documents/Projects/playground/j/node_modules/eslint/lib/util/source-code-fixer.js:76:26)
    at Function.SourceCodeFixer.applyFixes (/Users/gregholguin/Documents/Projects/playground/j/node_modules/eslint-plugin-html/src/index.js:152:23)
    at Linter.verifyAndFix (/Users/gregholguin/Documents/Projects/playground/j/node_modules/eslint/lib/linter.js:1258:43)
    at processText (/Users/gregholguin/Documents/Projects/playground/j/node_modules/eslint/lib/cli-engine.js:196:34)
    at processFile (/Users/gregholguin/Documents/Projects/playground/j/node_modules/eslint/lib/cli-engine.js:245:18)
    at executeOnFile (/Users/gregholguin/Documents/Projects/playground/j/node_modules/eslint/lib/cli-engine.js:585:25)
    at fileList.forEach.fileInfo (/Users/gregholguin/Documents/Projects/playground/j/node_modules/eslint/lib/cli-engine.js:622:13)
    at Array.forEach (<anonymous>)
    at CLIEngine.executeOnFiles (/Users/gregholguin/Documents/Projects/playground/j/node_modules/eslint/lib/cli-engine.js:621:18)
    at Object.execute (/Users/gregholguin/Documents/Projects/playground/j/node_modules/eslint/lib/cli.js:181:107)
```

**Comment by not-an-aardvark** [comment]
Thanks for the report. This looks like an issue with [`eslint-plugin-html`](https://github.com/BenoitZugmeyer/eslint-plugin-html) -- it was depending on the private `SourceCodeFixer` module used internally within ESLint, but that module was refactored in v4.6.0 and no longer exposes the same API.

**Comment by SimulatedGREG** [comment]
Thanks for the fast reply @not-an-aardvark ! I've made an issue over there.

**Comment by BenoitZugmeyer** [comment]
`eslint-plugin-html 3.2.1` supports ESLint 4.6+. This issue can be closed.

**Comment by not-an-aardvark** [comment]
Thanks @BenoitZugmeyer!

In an ideal world, I think it would be nice if external modules didn't rely on private APIs, since that would avoid issues like this and make builds more stable in general. I want to make ESLint's public API flexible enough that people don't need to monkeypatch, since there's a public API for what they want to do.

In this case, is there a public API that we could expose so that you wouldn't need to rely on monkeypatching `SourceCodeFixer`?

**Comment by BenoitZugmeyer** [comment]
Sure thing: if your [PR to allow autofix with plugins](https://github.com/eslint/eslint/pull/9090) lands, I probably won't need to patch `SourceCodeFixer` anymore. I didn't test it yet, but it looks good. I'll still need to patch `eslint.verify`, mainly to allow configuring file extensions to be considered as HTML via the `eslintrc` file. But this is another story, and your PR would make things simpler for me!
